lib/AST/Module.cpp

```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
   return nullptr;
 }
 
+//===----------------------------------------------------------------------===//
+// Cross-Import Overlays
+//===----------------------------------------------------------------------===//
+
+namespace swift {
+/// Represents a file containing information about cross-module overlays.
+class OverlayFile {
+  /// If not empty, contains the name of the cross-import file that must be
+  /// loaded; \c overlayModuleNames is not valid and must be empty.
+  ///
+  /// If empty, \c overlayModuleNames is valid and will contain the list of
+  /// module names unless loading the file failed.
+  StringRef filePath;
```

Comment:
I don't quite understand the logic here...

1. Why do we have an empty path when loading succeeds?
2. If loading the file fails, what state is this struct in? Is the path supposed to be empty and the vector is also empty?

---

There are three states we want to represent:

1. We have not yet read the file - Having the path here is necessary.
2. We tried reading the file and failed (I/O, deserialization etc.) - Having the path here is useful for debugging.
3. We tried reading the file and succeeded - Having the path here is potentially useful for debugging, in case we have trouble with loading the modules later on.

I think it would be simpler to have a dedicated 3-state enum instead of trying to squeeze the states into empty/non-empty combinations.

Reply:
This would also simplify the assertions in different places. We could have a single function (say `isValid`) which does the following checks:

1. The path is always non-empty.
2. The vector is empty if we have not yet read the file and non-empty if we did read the file successfully.

and use that check consistently.

Reply:
Okay, let's start with the preferred model. In Swift, I would probably define this type like this:

```swift
enum OverlayFile {
  case unloaded(filePath: String)
  case loaded(overlayNames: Set<Identifier>)
}
```

Could I add a third case for `cannotLoad`? Sure, but why would I? Once the error has been diagnosed, the rest of the compiler does not need to reason about the `OverlayFile.cannotLoad` case—it will behave exactly as though it were `OverlayFile.loaded(overlayNames: [])`. So why add this third case and make the rest of the code accommodate it?

-----

With that aside: the Swift model is a sum type, but I've actually implemented it with a product type. This is entirely because it would be a pain in the rear to write it as a sum type in our dialect of C++. We don't have `std::variant` in C++ 14. The values are not compatible with `llvm::PointerUnion`.

(Actually, it looks like we just added a TaggedUnion.h that could make this reasonable. Maybe I should just use that.)

Reply:
I agree with you that the minimal model that satisfies the requirements is the one you pointed out. The issue with the minimal model IMO is that it potentially makes debugging harder -- we are forgetting information (dropping the path, not keeping track of whether there was a load failure) even though it doesn't seem necessary to do that. It would be a different equation if we have thousands or hundreds of thousands of these, which is true for types, where we try to squeeze information into a low memory footprint.

If later down the line, say, we see a problem that we expected one cross-import overlay to be loaded but something else got loaded, debugging that forces us to look at the state in between (and how it is changing) when we have partial information instead of having the option to look at it in the end when we have maximum information.

---

> Once the error has been diagnosed, the rest of the compiler does not need to reason about the OverlayFile.cannotLoad case—it will behave exactly as though it were OverlayFile.loaded(overlayNames: []). So why add this third case and make the rest of the code accommodate it?

This is a good point. I was under the impression that not a lot of code was touching this directly but I haven't looked at the later commits (at 3/10 atm). If that's the case, do you think it might be useful to separate it out into two types an `OverlayFileWIP` and an `OverlayFile` with the latter only having the success state?

Reply:
I tried going full `TaggedUnion` to model this properly, but ran into some weird snarl involving Identifier's move constructor. I'll add a tristate.

Reply:
```
  enum class State { pending, loaded, failed };
  State state = State::pending;
```

Reply:
I've made a small comment about the naming convention `Pending` etc. (instead of `pending` etc.) but looks good now. 👍